Fix gathering inherited properties

BarkingBad opened the pull request:
Enables gathering super fields/properties by PSI's. This PR depends on model changes introduced in #2326 so I leave it as a draft (it is rebased onto that branch in order to build correctly. Nonetheless, it can be reviewed as a single commit.

vmishenev reviewed (approved):
I have reviewed only 122c6113998ee98c0018d9d2c8b1edccea1e4b8c here. 
Do you want to test `InheritedMember` in an extra?

IgnatBeresnev commented:
I've merged the changes this PR is based on, can you resolve conflicts please?

BarkingBad commented:
I've observed that the inhertited property is missing annotations, I will try to fix it that and include it in the tests

BarkingBad commented:
I know changes introduced in secondary commit are not directly coupled with the subject of the PR, but introduce small bugfixes that are related to the problem:
- minor changes with parameters of functions transforming kotlin to java (context to logger, since it is more accurate)
- fixed 2 smaller bugs with annotations and with names of accessors
- fixed issue with top-level `Property` TagWrapper for java beans functions that inherit description from the kotlin property, since it doesn't make any sense

I can part it to smaller PRs if demanded

BarkingBad commented on plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt:
```diff
@@ -647,13 +655,24 @@ private class DokkaDescriptorVisitor(
         return coroutineScope {
             val generics = async { descriptor.typeParameters.parallelMap { it.toVariantTypeParameter() } }
 
+            fun SourceSetDependent<DocumentationNode>.translateParamToDescription(): SourceSetDependent<DocumentationNode> {
+                return this.mapValues { (_, value) ->
+                    value.copy(children = value.children.map {
+                        when (it) {
+                            is Property -> Description(it.root)
+                            else -> it
+                        }
+                    })
+                }
+            }
```
It's a little hacky, though it makes no sense that inheriting getter/setter has `Property` wrapper from the description of the property in classlike kdoc header.

BarkingBad replied:
Done

BarkingBad commented on plugins/base/src/main/kotlin/translators/psi/DefaultPsiToDocumentableTranslator.kt:
```diff
@@ -452,7 +500,7 @@ class DefaultPsiToDocumentableTranslator(
                         when {
                             resolved.qualifiedName == "java.lang.Object" -> JavaObject(type.annotations())
-                            resolved is PsiTypeParameter -> getProjection(resolved)
+                            resolved is PsiTypeParameter -> getProjection(resolved).copy(extra = type.annotations())
```
It was actually a regression from the previous PR. The annotations are calculated from `type` PsiType, not inner `resolved` PsiTypeParameter

BarkingBad commented on plugins/base/src/main/kotlin/transformers/documentables/PropertiesMergerTransformer.kt:
```diff
@@ -0,0 +1,86 @@
+    } as T
+
+    private fun DFunction.getPropertyNameForFunction() =
+        when {
+            JvmAbi.isGetterName(name) -> propertyNameByGetMethodName(Name.identifier(name))?.asString()
+            JvmAbi.isSetterName(name) -> propertyNamesBySetMethodName(Name.identifier(name)).firstOrNull()
+                ?.asString()
+            else -> null
+        }
+
+    private fun mergePotentialBeansAndField(
+        functions: List<DFunction>,
+        fields: List<DProperty>
+    ): Pair<List<DFunction>, List<DProperty>> {
+        val fieldNames = fields.associateBy { it.name }
+        val accessors = mutableMapOf<DProperty, MutableList<DFunction>>()
+        val regularMethods = mutableListOf<DFunction>()
+        functions.forEach { method ->
+            val field = method.getPropertyNameForFunction()?.let { name -> fieldNames[name] }
+            if (field != null) {
+                accessors.getOrPut(field, ::mutableListOf).add(method)
+            } else {
+                regularMethods.add(method)
+            }
+        }
+        return regularMethods.toList() to accessors.map { (dProperty, dFunctions) ->
+            if (dProperty.visibility.values.all { it is KotlinVisibility.Private }) {
+                dFunctions.flatMap { it.visibility.values }.toSet().singleOrNull()?.takeIf {
+                    it in listOf(KotlinVisibility.Public, KotlinVisibility.Protected)
+                }?.let { visibility ->
+                    dProperty.copy(
+                        getter = dFunctions.firstOrNull { it.type == dProperty.type },
+                        setter = dFunctions.firstOrNull { it.parameters.isNotEmpty() },
+                        visibility = dProperty.visibility.mapValues { visibility }
+                    )
+                } ?: dProperty
+            } else {
+                dProperty
+            }
+        } + fields.toSet().minus(accessors.keys.toSet())
```
This is more or less copied from PsiToDocumentableTranslator. I wonder if we could remove it from there since this transformer should work well for PSIs and Descriptor sourced Documentables. However, I don't have time to experiment with that :(

BarkingBad replied:
Lines 581 - 603: functions `getPropertyNameForFunction` and `splitFunctionsAndAccessors`

These are losely copied and tailored to our needs here, yet it was copy-pasted. I will annotate it with todo as you asked

BarkingBad commented on plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt:
```diff
@@ -430,7 +430,10 @@ private class DokkaDescriptorVisitor(
         parent: DRIWithPlatformInfo
     ): DProperty {
-        val (dri, inheritedFrom) = originalDescriptor.createDRI()
+        val (dri, inheritedFrom) = originalDescriptor.createDRI().let {
+            if (it.second != null) (it.second to it.first) as Pair<DRI, DRI?>
+            else it
+        }
```
Most tricky part. I believe this change was a little https://github.com/Kotlin/dokka/commit/37243ad7a5152d4f0f30bb8f1446d4d5c6ce2094 was not so accurate (In case of B extending A, I was getting `DRI` for A class and `inheritedFrom` for B class.)

BarkingBad replied:
Actually, my implementation was also wrong, see newer commit

BarkingBad commented on plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt:
```diff
@@ -485,7 +488,10 @@ private class DokkaDescriptorVisitor(
         parent: DRIWithPlatformInfo
     ): DFunction {
-        val (dri, inheritedFrom) = originalDescriptor.createDRI()
+        val (dri, inheritedFrom) = originalDescriptor.createDRI().let {
+            if (it.second != null) (it.second to it.first) as Pair<DRI, DRI?>
+            else it
+        }
```
Same as above

BarkingBad commented on plugins/base/src/test/kotlin/model/PropertyTest.kt:
```diff
@@ -152,12 +155,15 @@ class PropertyTest : AbstractModelTest("/src/main/kotlin/property/Test.kt", "pro
             with((this / "property").cast<DPackage>()) {
                 with((this / "Bar" / "property").cast<DProperty>()) {
-                    dri.classNames equals "Foo"
+                    dri.classNames equals "Bar"
                     name equals "property"
                     children counts 0
                     with(getter.assertNotNull("Getter")) {
                         type.name equals "Int"
                     }
+                    extra[InheritedMember]?.inheritedFrom?.values?.single()?.run {
+                        classNames equals "Foo"
+                    }
```
I believe this test was wrong, it was correct in the context of previous implementation, but we expect property to have its class in the DRI and actual class it is coming from in the InheritedMember extra property, am I right?

BarkingBad commented on plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt:
```diff
@@ -430,10 +430,8 @@ private class DokkaDescriptorVisitor(
-            if (it.second != null) (it.second to it.first) as Pair<DRI, DRI?>
-            else it
-        }
+        val (dri, overridenBy) = originalDescriptor.createDRI()
+        val inheritedFrom = dri.copy(callable = null).takeIf { parent.dri.classNames != dri.classNames }
```
`createDRI` returns the DRI of the exact element and potential DRI of an element that is overriding it (It can be also FAKE_OVERRIDE which is in fact just inheritance of the symbol)

Looking at what PSIs do, they give the DRI of the element within the classnames where it is actually declared and inheritedFrom as the same DRI but truncated callable part, I think it is wise to unify that. Nonetheless, the second output of the tuple was never correct `inheritedFrom` value and has to be changed.

Now I think I need to double-check the dri if it is not a fake override, but an actual override. Also, I was thinking whether we could rely on existence `overridenBy` as the guard for `inheritedFrom` creation.

IgnatBeresnev commented on plugins/base/src/test/kotlin/superFields/PsiSuperFieldsTest.kt:
```diff
@@ -0,0 +1,142 @@
+
+class PsiSuperFieldsTest : BaseAbstractTest() {
+
+    @Disabled // TODO: Remove with Kotlin 1.6.20
```
We've updated to 1.6.20-M1, but not everything that will go into the release is in M1, I think. You think this can be removed now? Or are we waiting for some specific task to be released in the compiler?

BarkingBad replied:
We should check if tests are passing without flakiness

BarkingBad replied:
I've checked that and it still keeps failing sometimes. We should keep them disabled

IgnatBeresnev commented on plugins/javadoc/src/test/kotlin/org/jetbrains/dokka/javadoc/JavadocIndexTest.kt:
```diff
@@ -21,7 +21,7 @@ internal class JavadocIndexTest : AbstractJavadocTemplateMapTest() {
 
         testIndexPages { indexPages ->
-            assertEquals(if (hasAdditionalFunction()) 41 else 40, indexPages.sumBy { it.elements.size })
+            assertEquals(if (hasAdditionalFunction()) 32 else 31, indexPages.sumBy { it.elements.size })
```
Out of curiosity, why such a drop in numbers? Where did 10 functions go? :))

BarkingBad replied:
These are synthetic methods (toString, etc.) and they got merged, earlier, the `inheritedFrom` DRI was differing because it was not always pointing to Object/Any

IgnatBeresnev commented on plugins/kotlin-as-java/src/main/kotlin/transformToJava.kt:
```diff
@@ -7,23 +7,24 @@ import org.jetbrains.dokka.model.DFunction
 }
 
-fun DPackage.transformToJava(context: DokkaContext): DPackage {
-    return JVM_NAME_DOCUMENTABLE_TRANSFORMER.transform(this.asJava(), context)
+fun DPackage.transformToJava(logger: DokkaLogger): DPackage {
+    return JVM_NAME_DOCUMENTABLE_TRANSFORMER.transform(this.asJava(), logger)
```
Any particular reason for changing context to just logger? Since this is public API, maybe it's worth having `context` here? It'll take some deprication cycles to change this back to context in the future

IgnatBeresnev commented on plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt:
```diff
@@ -430,7 +430,8 @@ private class DokkaDescriptorVisitor(
         parent: DRIWithPlatformInfo
     ): DProperty {
-        val (dri, inheritedFrom) = originalDescriptor.createDRI()
+        val (dri, _) = originalDescriptor.createDRI()
+        val inheritedFrom = dri.copy(callable = null).takeIf { parent.dri.classNames != dri.classNames }
```
Oh, it seems you've already answered that, but the pr comment was marked "outdated"

It would be really helpful to document some of this in code

> createDRI returns the DRI of the exact element and potential DRI of an element that is overriding it (It can be also FAKE_OVERRIDE which is in fact just inheritance of the symbol)
> 
> Looking at what PSIs do, they give the DRI of the element within the classnames where it is actually declared and inheritedFrom as the same DRI but truncated callable part, I think it is wise to unify that. Nonetheless, the second output of the tuple was never correct inheritedFrom value and has to be changed.
> 
> Now I think I need to double-check the dri if it is not a fake override, but an actual override. Also, I was thinking whether we could rely on existence overridenBy as the guard for inheritedFrom creation.

BarkingBad replied:
Done

BarkingBad commented on plugins/base/src/main/kotlin/transformers/documentables/PropertiesMergerTransformer.kt:
```diff
@@ -0,0 +1,86 @@
+            else -> null
+        }
+
+    private fun mergePotentialBeansAndField(
```
It turns out I wrongly interpreted what JavaBean is (https://stackoverflow.com/a/3295517). I thought bean referred to getters/setters convention, but apparently, it is describing the whole class. Will rename then

BarkingBad commented on plugins/base/src/main/kotlin/translators/psi/DefaultPsiToDocumentableTranslator.kt:
```diff
@@ -431,6 +470,15 @@ class DefaultPsiToDocumentableTranslator(
+        /**
+         * Workaround for getting JvmField Kotlin annotation in PSIs
+         */
+        private fun Collection<PsiAnnotation>.getJvmFieldAnnotation() = filter {
+            it.qualifiedName == "kotlin.jvm.JvmField"
+        }.map {
+            Annotations.Annotation(DRI("kotlin.jvm", "JvmField"), emptyMap())
+        }.distinct()
```
I will look into that

BarkingBad replied:
Done

BarkingBad commented on plugins/base/src/main/kotlin/transformers/documentables/PropertiesMergerTransformer.kt:
```diff
@@ -0,0 +1,86 @@
+                regularMethods.add(method)
+            }
+        }
+        return regularMethods.toList() to accessors.map { (dProperty, dFunctions) ->
```
Ok, will try to debunk it

BarkingBad replied:
Done

BarkingBad commented on plugins/base/src/main/kotlin/transformers/documentables/PropertiesMergerTransformer.kt:
```diff
@@ -0,0 +1,86 @@
+        return regularMethods.toList() to accessors.map { (dProperty, dFunctions) ->
+            if (dProperty.visibility.values.all { it is KotlinVisibility.Private }) {
+                dFunctions.flatMap { it.visibility.values }.toSet().singleOrNull()?.takeIf {
+                    it in listOf(KotlinVisibility.Public, KotlinVisibility.Protected)
```
Will look into that

BarkingBad replied:
I think we should hardcode the configurations like these manually... I don't know if I have time to do it this week.

BarkingBad commented on plugins/base/src/main/kotlin/transformers/documentables/PropertiesMergerTransformer.kt:
```diff
@@ -0,0 +1,86 @@
+import org.jetbrains.kotlin.load.java.propertyNamesBySetMethodName
+import org.jetbrains.kotlin.name.Name
+
+class PropertiesMergerTransformer : PreMergeDocumentableTransformer {
```
Done

BarkingBad commented on plugins/base/src/main/kotlin/transformers/documentables/PropertiesMergerTransformer.kt:
```diff
@@ -0,0 +1,86 @@
+
+    private fun <T : Documentable> T.mergeBeansAndField(): T = when (this) {
+        is DClass -> {
+            val (functions, properties) = mergePotentialBeansAndField(this.functions, this.properties)
```
Done

IgnatBeresnev commented on plugins/base/src/main/kotlin/transformers/documentables/PropertiesMergerTransformer.kt:
```diff
@@ -0,0 +1,118 @@
+
+/**
+ * This transformer is used to merge the backing fields and accessors (getters and setters)
+ * obtained from Java sources. This way, we could generate more coherent documentation,
```
Is this transformer only applicable to java sources? If so:
1. what do you think about renaming it to `JavaPropertiesMergerTransformer`? 
2. Can we somehow filter processed items and avoid processing kotlin classes? Does it make sense? Although I don't know if we have the ability to extract that info from DClass. Or maybe you already have that filter somewhere, e.g via `KotlinVisibility`?

My main concern is that if we don't limit it, then kotlin classes might satisfy the same conditions as Java and it'll introduce bugs (i.e not display getters/setters of a kotlin class for explicit declarations). If it's only java classes that are processed (somehow), should be safer. But maybe this can never happen?

BarkingBad replied:
Technically it can happen, I am not sure how I could limit it to only Java classes. Do you have any idea?

BarkingBad commented:
I'll take a look tomorrow on that

vmishenev reviewed (changes requested):
I have discovered the same logic to merge accessors with fields in the different places: `PropertiesMergerTransformer` and `DefaultPsiToDocumentableTranslator`. What do you about extraction this logic from  `DefaultPsiToDocumentableTranslator` into `PropertiesMergerTransformer`?

BarkingBad commented on plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt:
```diff
@@ -647,13 +662,34 @@ private class DokkaDescriptorVisitor(
+             * some opens-source plugin creators.
+             * TODO: Should rethink if we could fix it globally in dokka or in compiler itself.
+             */
+            fun SourceSetDependent<DocumentationNode>.translatePropertyToDescription(): SourceSetDependent<DocumentationNode> {
```
Yes, that's right

BarkingBad replied:
Should I rename it?

BarkingBad commented on plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt:
```diff
@@ -647,13 +662,34 @@ private class DokkaDescriptorVisitor(
+             *
+             * Actually, the problem impacts more of these tags, yet this particular tag was blocker for
+             * some opens-source plugin creators.
+             * TODO: Should rethink if we could fix it globally in dokka or in compiler itself.
```
It could be fixed in the compiler. On the other hand, the compiler returns corresponding KDoc for some property etc. that is based on the positions of the original KDoc string, and therefore it is wrapped in the `Property` section, which is somehow correct. There could be some service in dokka to translate all of these, though one should rethink all the possibilities and cover them manually. Probably it won't be so hard afterall

BarkingBad commented on plugins/base/src/test/kotlin/superFields/PsiSuperFieldsTest.kt:
```diff
@@ -0,0 +1,142 @@
+                    assertNull(this.getter)
+                    assertNull(this.setter)
+                    assertNotNull(this.extra[Annotations]?.directAnnotations?.values?.single()?.find {
+                        it.dri.packageName == "kotlin.jvm" &&
```
Done

BarkingBad commented:
> I have discovered the same logic to merge accessors with fields in the different places: PropertiesMergerTransformer and DefaultPsiToDocumentableTranslator. What do you about extraction this logic from DefaultPsiToDocumentableTranslator into PropertiesMergerTransformer?

I honestly don't remember why I copied it (I know it is redundant, I warned about that), but don't remember the case when the properties won't be merged in PSI translator already.

vmishenev commented on plugins/base/src/main/kotlin/translators/descriptors/DefaultDescriptorToDocumentableTranslator.kt:
```diff
@@ -430,7 +430,16 @@ private class DokkaDescriptorVisitor(
+         * declared and inheritedFrom as the same DRI but truncated callable part.
+         * Therefore, we set callable to null and take the DRI only if it is indeed coming from different class.
+         */
+        val inheritedFrom = dri.copy(callable = null).takeIf { parent.dri.classNames != dri.classNames }
```
Yes, it should.

vmishenev commented:
So are there some blockers to merge accessors with fields only in `PropertiesMergerTransformer`?
I do not like we have the same code in the different places.
I suggested keeping all the fields and methods  in `DefaultPsiToDocumentableTranslator` as is. Then  `PropertiesMergerTransformer` can process them to merge accessors with properties.  

> It also answers @IgnatBeresnev question about limiting this transformer to only Java sources.

We need to merge properties only inside a Java class. Is it right? So is it possible to limit the transformer to only Java class?

IgnatBeresnev commented:
Hi!

The PR is starting to rot a little, there are some conflicts and new build errors. I've checked out your changes locally and rebased onto latest master, and I'm planning to remove duplicated code as we discussed above -- I know it's hard to find time for this, so we don't want to burden you too much with this :)

I'm closing this PR to preserve history (don't want to force-push over your commits and comments), but your changes have migrated to https://github.com/Kotlin/dokka/pull/2481 and will be merged after some additional changes. Thanks for your work!
